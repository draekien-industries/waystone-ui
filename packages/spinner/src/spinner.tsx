import { getSpinnerCss } from './spinner.styles';
import type { SpinnerProps } from './spinner.types';

export const Spinner = (props: SpinnerProps) => (
  <div sx={getSpinnerCss(props)} aria-description="Loading..." />
);
