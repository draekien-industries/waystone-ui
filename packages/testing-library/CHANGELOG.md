# @waystone/testing-library

## 9.2.0

### Minor Changes

- 21db598: chore: bump deps and add documentation

### Patch Changes

- Updated dependencies [21db598]
  - @waystone/theme@9.3.0

## 9.1.6

### Patch Changes

- 41664e2: fix: use type imports to improve bundling
- Updated dependencies [41664e2]
  - @waystone/theme@9.2.5

## 9.1.5

### Patch Changes

- 856ba2b: fix: missing dependencies when installing package
- Updated dependencies [856ba2b]
  - @waystone/theme@9.2.4

## 9.1.4

### Patch Changes

- 5b726aa: fix build and dependencies
- Updated dependencies [5b726aa]
  - @waystone/theme@9.2.3

## 9.1.3

### Patch Changes

- 7086019: Update dependencies
- Updated dependencies [7086019]
  - @waystone/theme@9.2.2

## 9.1.2

### Patch Changes

- @waystone/theme@9.2.1

## 9.1.1

### Patch Changes

- Updated dependencies [8f75dd6]
  - @waystone/theme@9.2.0

## 9.1.0

### Minor Changes

- 5c43357: Switch to pnpm and implement minimal syntax highlighter

### Patch Changes

- Updated dependencies [5c43357]
  - @waystone/theme@9.1.0

## 9.0.0

### Major Changes

- 4a19c42: Update package structure

  Created the `@waystone/ui` package for easier installation and created the `@waystone/testing-library`
  package for a preconfigured testing-library render function.

  Uninstall your `@waystone/*` packages, and run the command `yarn add @waystone/ui`
  to update to the latest version. You will need to update the imports to point
  to the new source.

### Patch Changes

- Updated dependencies [4a19c42]
  - @waystone/theme@9.0.0
