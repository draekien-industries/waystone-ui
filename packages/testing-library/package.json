{
  "name": "@waystone/testing-library",
  "version": "10.0.2",
  "private": true,
  "main": "./src/index.ts",
  "repository": {
    "type": "git",
    "url": "https://github.com/draekien-industries/waystone-ui"
  },
  "description": "The react component library used by draekien-industries",
  "author": "draekien-industries",
  "sideEffects": false,
  "license": "MIT",
  "files": [
    "src/**"
  ],
  "scripts": {
    "lint": "eslint --ext .tsx,.ts,.jsx,.js ./src/ --fix",
    "lint:release": "eslint --ext .tsx,.ts,.jsx,.js ./src/"
  },
  "devDependencies": {
    "@waystone/tsconfig": "workspace:*",
    "eslint-config-waystone": "workspace:*",
    "eslint-plugin-waystone": "workspace:*",
    "react": "^18.3.1",
    "react-dom": "^18.3.1"
  },
  "dependencies": {
    "@testing-library/jest-dom": "^6.6.3",
    "@testing-library/react": "^16.2.0",
    "@testing-library/user-event": "^14.6.0",
    "@theme-ui/color": "^0.17.1",
    "@theme-ui/css": "^0.17.1",
    "@waystone/theme-provider": "workspace:*",
    "jest": "^29.7.0",
    "jest-environment-jsdom": "^29.7.0",
    "theme-ui": "^0.17.1"
  },
  "publishConfig": {
    "access": "public"
  }
}
