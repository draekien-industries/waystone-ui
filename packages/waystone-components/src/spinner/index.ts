export * from './spinner';
export * from './spinner.hoc';
export * from './spinner.hooks';
export * from './spinner.context';
