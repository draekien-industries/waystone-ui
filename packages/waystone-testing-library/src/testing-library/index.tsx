'use client';

import type { RenderOptions, RenderResult } from '@testing-library/react';
import { render } from '@testing-library/react';
import { Waystone } from '@waystone/theme-provider';
import type { PropsWithChildren } from 'react';
import type React from 'react';
import { useEffect } from 'react';
import { useColorMode } from 'theme-ui';

const TestProvider = ({ children }: PropsWithChildren) => (
  <Waystone>{children}</Waystone>
);

const DarkModeWrapper = ({ children }: PropsWithChildren) => {
  const [, setColorMode] = useColorMode();

  useEffect(() => {
    setColorMode('dark');
  }, [setColorMode]);

  return <div>{children}</div>;
};

const themedRender: (
  ui: React.ReactElement,
  options?: RenderOptions
) => RenderResult = (ui, options?) =>
  render(ui, { wrapper: options?.wrapper || TestProvider, ...options });

export * from '@testing-library/react';
export * from '@testing-library/user-event';

export { DarkModeWrapper, themedRender as render };
