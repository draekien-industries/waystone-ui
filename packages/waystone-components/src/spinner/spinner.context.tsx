import React, { PropsWithChildren } from 'react';
import { useSpinner } from './spinner.hooks';

/** The value provided by the {@link SpinnerContext} */
export interface SpinnerContextValue {
  /** A function that shows the spinner. */
  show?: () => void;
  /** A function that hides the spinner. */
  hide?: () => void;
}

export interface SpinnerProviderProps {
  initialVisible?: boolean;
}

/** The context for the {@link SpinnerContextValue} */
export const SpinnerContext = React.createContext<SpinnerContextValue>({});

/**
 * A component that provides the {@link SpinnerContextValue} to its children
 * and renders the spinner component onto the DOM.
 * @param props.children - the children of the component
 * @returns the spinner context provider
 */
export const SpinnerProvider: React.FC<
  PropsWithChildren<SpinnerProviderProps>
> = ({ initialVisible = true, children }) => {
  const { show, hide, spinner } = useSpinner({ initialVisible });
  const value = React.useMemo(() => ({ show, hide }), [show, hide]);

  return (
    <SpinnerContext.Provider value={value}>
      {children}
      {spinner}
    </SpinnerContext.Provider>
  );
};

/**
 * A hook that provides access to the {@link SpinnerContextValue} when
 * used within a {@link SpinnerProvider}.
 * @returns the spinner context value
 */
export const useSpinnerContext = () => {
  const context = React.useContext(SpinnerContext);
  if (!context) {
    throw new Error(
      '"useSpinnerContext" must be used within a "SpinnerProvider"'
    );
  }

  return context;
};
