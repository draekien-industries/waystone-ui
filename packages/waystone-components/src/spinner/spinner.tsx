import { Overlay } from '../overlay/overlay';
import type { SpinnerCssProps } from './spinner.styles';
import { spinnerCss } from './spinner.styles';

export type SpinnerProps = {
  /**
   * Renders the spinner with an overlay that takes up the full
   * width and height of the parent container.
   */
  overlay?: boolean;
} & SpinnerCssProps;

/** A terminal style loading spinner */
export const Spinner = ({
  size = 'medium',
  color = 'accent',
  fullWidth = false,
  overlay = false,
}: SpinnerProps) => {
  if (overlay) {
    return (
      <Overlay>
        <span
          sx={{
            ...spinnerCss({ size, color, fullWidth: true }),
            height: '100%',
            zIndex: 'highest',
            position: 'absolute',
          }}
        />
      </Overlay>
    );
  }

  return (
    <span
      role="progressbar"
      sx={spinnerCss({ size, color, fullWidth })}
      aria-valuetext="Loading..."
      aria-busy
    />
  );
};
