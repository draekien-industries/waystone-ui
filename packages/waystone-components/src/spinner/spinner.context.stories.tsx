import React from 'react';
import { ComponentMeta, ComponentStory } from '@storybook/react';
import { within, waitFor, userEvent } from '@storybook/testing-library';
import { UseSpinnerContextDemo } from './spinner.internal';
import { SpinnerProvider } from './spinner.context';

export default {
  title: 'Hooks/useSpinnerContext',
  component: UseSpinnerContextDemo,
  args: {
    initialVisible: true,
  },
} as ComponentMeta<typeof UseSpinnerContextDemo>;

const Template: ComponentStory<typeof UseSpinnerContextDemo> = (args) => (
  <SpinnerProvider {...args}>
    <UseSpinnerContextDemo />
  </SpinnerProvider>
);

export const Default = Template.bind({});

Default.play = async ({ canvasElement }) => {
  const canvas = within(canvasElement);

  await waitFor(() =>
    setTimeout(async () => {
      await userEvent.click(canvas.getByText('Hide'));
    }, 5000)
  );
};
