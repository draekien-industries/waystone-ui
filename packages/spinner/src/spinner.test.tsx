import { render } from '@waystone/testing-library';
import { Spinner } from './spinner';
import type { SpinnerSize } from './spinner.types';
import {
  borderCss,
  getSpinnerCss,
  pathKeyframes,
  transformKeyframes,
  widthCss,
} from './spinner.styles';

describe('<Spinner />', () => {
  const sizes = ['small', 'medium', 'large'] as const satisfies SpinnerSize[];

  describe('component', () => {
    describe.each(sizes)('size %s', (size) => {
      it('should match snapshot', () => {
        const { container } = render(<Spinner size={size} />);

        expect(container).toMatchSnapshot();
      });
    });
  });

  describe('css', () => {
    describe('pathKeyframes', () => {
      it('should match snapshot', () => {
        expect(pathKeyframes).toMatchSnapshot();
      });
    });

    describe('transformKeyframes', () => {
      it('should match snapshot', () => {
        expect(transformKeyframes).toMatchSnapshot();
      });
    });

    describe('widthCss', () => {
      it('should match snapshot', () => {
        expect(widthCss).toMatchSnapshot();
      });
    });

    describe('borderCss', () => {
      it('should match snapshot', () => {
        expect(borderCss).toMatchSnapshot();
      });
    });

    describe.each(sizes)('size %s', (size) => {
      it('should match snapshot', () => {
        expect(getSpinnerCss({ size })).toMatchSnapshot();
      });
    });
  });
});
