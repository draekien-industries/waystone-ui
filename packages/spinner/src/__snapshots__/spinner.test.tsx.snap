// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Spinner /> component size large should match snapshot 1`] = `
<div>
  <div
    aria-description="Loading..."
    class="css-w8xjas-Spinner"
  />
</div>
`;

exports[`<Spinner /> component size medium should match snapshot 1`] = `
<div>
  <div
    aria-description="Loading..."
    class="css-c1ojwv-Spinner"
  />
</div>
`;

exports[`<Spinner /> component size small should match snapshot 1`] = `
<div>
  <div
    aria-description="Loading..."
    class="css-22drur-Spinner"
  />
</div>
`;

exports[`<Spinner /> css borderCss should match snapshot 1`] = `
{
  "large": "0.5rem solid",
  "medium": "0.25rem solid",
  "small": "0.25rem solid",
}
`;

exports[`<Spinner /> css pathKeyframes should match snapshot 1`] = `
{
  "anim": 1,
  "name": "animation-18cc70e",
  "styles": "@keyframes animation-18cc70e{0%{clip-path:polygon(50% 50%, 0 0, 50% 0%, 50% 0%, 50% 0%, 50% 0%, 50% 0% );}12.5%{clip-path:polygon(50% 50%, 0 0, 50% 0%, 100% 0%, 100% 0%, 100% 0%, 100% 0% );}25%{clip-path:polygon(50% 50%, 0 0, 50% 0%, 100% 0%, 100% 100%, 100% 100%, 100% 100% );}50%{clip-path:polygon(50% 50%, 0 0, 50% 0%, 100% 0%, 100% 100%, 50% 100%, 0% 100% );}62.5%{clip-path:polygon(50% 50%, 100% 0, 100% 0%, 100% 0%, 100% 100%, 50% 100%, 0% 100% );}75%{clip-path:polygon(50% 50%, 100% 100%, 100% 100%, 100% 100%, 100% 100%, 50% 100%, 0% 100% );}100%{clip-path:polygon(50% 50%, 50% 100%, 50% 100%, 50% 100%, 50% 100%, 50% 100%, 0% 100% );}}",
  "toString": [Function],
}
`;

exports[`<Spinner /> css size large should match snapshot 1`] = `
{
  "animation": "_EMO_animation-18cc70e_@keyframes animation-18cc70e{0%{clip-path:polygon(50% 50%, 0 0, 50% 0%, 50% 0%, 50% 0%, 50% 0%, 50% 0% );}12.5%{clip-path:polygon(50% 50%, 0 0, 50% 0%, 100% 0%, 100% 0%, 100% 0%, 100% 0% );}25%{clip-path:polygon(50% 50%, 0 0, 50% 0%, 100% 0%, 100% 100%, 100% 100%, 100% 100% );}50%{clip-path:polygon(50% 50%, 0 0, 50% 0%, 100% 0%, 100% 100%, 50% 100%, 0% 100% );}62.5%{clip-path:polygon(50% 50%, 100% 0, 100% 0%, 100% 0%, 100% 100%, 50% 100%, 0% 100% );}75%{clip-path:polygon(50% 50%, 100% 100%, 100% 100%, 100% 100%, 100% 100%, 50% 100%, 0% 100% );}100%{clip-path:polygon(50% 50%, 50% 100%, 50% 100%, 50% 100%, 50% 100%, 50% 100%, 0% 100% );}}_EMO_ 0.8s infinite linear alternate, _EMO_animation-y2w2h2_@keyframes animation-y2w2h2{0%{transform:scaleY(1) rotate(0deg);}49.99%{transform:scaleY(1) rotate(135deg);}50%{transform:scaleY(-1) rotate(0deg);}100%{transform:scaleY(-1) rotate(-135deg);}}_EMO_ 1.6s infinite linear",
  "aspectRatio": 1,
  "border": "0.5rem solid",
  "borderColor": "highlight",
  "borderRadius": "50%",
  "width": "3rem",
}
`;

exports[`<Spinner /> css size medium should match snapshot 1`] = `
{
  "animation": "_EMO_animation-18cc70e_@keyframes animation-18cc70e{0%{clip-path:polygon(50% 50%, 0 0, 50% 0%, 50% 0%, 50% 0%, 50% 0%, 50% 0% );}12.5%{clip-path:polygon(50% 50%, 0 0, 50% 0%, 100% 0%, 100% 0%, 100% 0%, 100% 0% );}25%{clip-path:polygon(50% 50%, 0 0, 50% 0%, 100% 0%, 100% 100%, 100% 100%, 100% 100% );}50%{clip-path:polygon(50% 50%, 0 0, 50% 0%, 100% 0%, 100% 100%, 50% 100%, 0% 100% );}62.5%{clip-path:polygon(50% 50%, 100% 0, 100% 0%, 100% 0%, 100% 100%, 50% 100%, 0% 100% );}75%{clip-path:polygon(50% 50%, 100% 100%, 100% 100%, 100% 100%, 100% 100%, 50% 100%, 0% 100% );}100%{clip-path:polygon(50% 50%, 50% 100%, 50% 100%, 50% 100%, 50% 100%, 50% 100%, 0% 100% );}}_EMO_ 0.8s infinite linear alternate, _EMO_animation-y2w2h2_@keyframes animation-y2w2h2{0%{transform:scaleY(1) rotate(0deg);}49.99%{transform:scaleY(1) rotate(135deg);}50%{transform:scaleY(-1) rotate(0deg);}100%{transform:scaleY(-1) rotate(-135deg);}}_EMO_ 1.6s infinite linear",
  "aspectRatio": 1,
  "border": "0.25rem solid",
  "borderColor": "highlight",
  "borderRadius": "50%",
  "width": "1.5rem",
}
`;

exports[`<Spinner /> css size small should match snapshot 1`] = `
{
  "animation": "_EMO_animation-18cc70e_@keyframes animation-18cc70e{0%{clip-path:polygon(50% 50%, 0 0, 50% 0%, 50% 0%, 50% 0%, 50% 0%, 50% 0% );}12.5%{clip-path:polygon(50% 50%, 0 0, 50% 0%, 100% 0%, 100% 0%, 100% 0%, 100% 0% );}25%{clip-path:polygon(50% 50%, 0 0, 50% 0%, 100% 0%, 100% 100%, 100% 100%, 100% 100% );}50%{clip-path:polygon(50% 50%, 0 0, 50% 0%, 100% 0%, 100% 100%, 50% 100%, 0% 100% );}62.5%{clip-path:polygon(50% 50%, 100% 0, 100% 0%, 100% 0%, 100% 100%, 50% 100%, 0% 100% );}75%{clip-path:polygon(50% 50%, 100% 100%, 100% 100%, 100% 100%, 100% 100%, 50% 100%, 0% 100% );}100%{clip-path:polygon(50% 50%, 50% 100%, 50% 100%, 50% 100%, 50% 100%, 50% 100%, 0% 100% );}}_EMO_ 0.8s infinite linear alternate, _EMO_animation-y2w2h2_@keyframes animation-y2w2h2{0%{transform:scaleY(1) rotate(0deg);}49.99%{transform:scaleY(1) rotate(135deg);}50%{transform:scaleY(-1) rotate(0deg);}100%{transform:scaleY(-1) rotate(-135deg);}}_EMO_ 1.6s infinite linear",
  "aspectRatio": 1,
  "border": "0.25rem solid",
  "borderColor": "highlight",
  "borderRadius": "50%",
  "width": "1rem",
}
`;

exports[`<Spinner /> css transformKeyframes should match snapshot 1`] = `
{
  "anim": 1,
  "name": "animation-y2w2h2",
  "styles": "@keyframes animation-y2w2h2{0%{transform:scaleY(1) rotate(0deg);}49.99%{transform:scaleY(1) rotate(135deg);}50%{transform:scaleY(-1) rotate(0deg);}100%{transform:scaleY(-1) rotate(-135deg);}}",
  "toString": [Function],
}
`;

exports[`<Spinner /> css widthCss should match snapshot 1`] = `
{
  "large": "3rem",
  "medium": "1.5rem",
  "small": "1rem",
}
`;
